app/src/main/java/io/appium/settings/RecorderThread.java

```diff
@@ -0,0 +1,367 @@
+                if (encoderStatus == MediaCodec.INFO_TRY_AGAIN_LATER) {
+                    Log.w(TAG, "encoder.dequeueOutputBuffer: try again");
+                } else if (encoderStatus == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
+                    if (videoTrackIndex > 0) {
+                        Log.e(TAG, "videoTrackIndex less than zero");
+                        break;
+                    }
+                    videoTrackIndex = muxer.addTrack(videoEncoder.getOutputFormat());
+                    startMuxerIfSetUp();
+                } else if (encoderStatus < 0) {
+                    Log.w(TAG, "unexpected result from encoder.dequeueOutputBuffer: "
+                            + encoderStatus);
+                } else {
+                    ByteBuffer encodedData = videoEncoder.getOutputBuffer(encoderStatus);
+                    if (encodedData == null) {
+                        Log.w(TAG, "videoEncoder, encodedData null");
+                        break;
+                    }
+
+                    if (bufferInfo.size != 0 &&
+                            (bufferInfo.flags & MediaCodec.BUFFER_FLAG_CODEC_CONFIG) == 0) {
+                        bufferInfo.presentationTimeUs = getPresentationTimeUs();
+                        muxer.writeSampleData(videoTrackIndex, encodedData, bufferInfo);
+                    }
+
+                    videoEncoder.releaseOutputBuffer(encoderStatus, false);
+
+                    if ((bufferInfo.flags & MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0) {
+                        Log.w(TAG, "videoEncoder, buffer eos");
+                        break;
+                    }
+                }
+            }
+        } catch (Exception mainException) {
+            mainException.printStackTrace();
+        } finally {
+            try {
+                if (muxer != null) {
+                    muxer.stop();
+                    muxer.release();
+                    muxer = null;
+                }
+
+                if (virtualDisplay != null) {
+                    virtualDisplay.release();
+                    virtualDisplay = null;
+                }
+
+                if (surface != null) {
+                    surface.release();
+                    surface = null;
+                }
+
+                if (videoEncoder != null) {
+                    videoEncoder.stop();
+                    videoEncoder.release();
+                    videoEncoder = null;
+                }
+
+                if (audioRecordThread != null) {
+                    audioStopped = true;
+                    audioRecordThread.join();
+                }
+
+                if (audioEncoder != null) {
+                    audioEncoder.stop();
+                    audioEncoder.release();
+                    audioEncoder = null;
+                }
+            } catch (Exception e) {
+                Log.w(TAG, "Error releasing resources", e);
```

Comment:
this might not work as expected if an exception happens, because then all further resources won't be released properly

Reply:
> Thank you. Is there any reason why you did not add this feature in https://github.com/appium/appium-uiautomator2-server , for example, as a `mobile:` command? https://github.com/appium/appium-uiautomator2-driver (kind of a basic question as my out of curiosity)
> 
> [updated]
> 
> ah, this was for the below comment in the linked issue?
> 
> > Which means one can simply install any sound recording app, control it using UIA2 to start/stop recording process and be happy about it %)

yeah since recording on android via public android APIs are only available on application level, thats why i targeted io.appium.settings app and for future direction, i think yeah the start/stop recording commands can be folded into much less lenghty commands like `mobile: startRecordingEtc.`

Reply:
done, my bad sorry, separated try-catches